### PR TITLE
add more languages

### PR DESCRIPTION
**Pull request by StefanVukovic99**
On top of #584, adds more languages and their text preprocessors. Minor changes to the pattern set there with the types/organization.

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -502,14 +502,36 @@
                 "ext/js/general/object-property-accessor.js",
                 "ext/js/general/regex-util.js",
                 "ext/js/general/text-source-map.js",
+                "ext/js/language/ar/language-arabic.js",
```
Comment (StefanVukovic99) [inline .eslintrc.json:505]: I think using wildcards like `language/**/*.js` would be better here but couldn't get the test to pass.

File: types/ext/language.d.ts [hunk 2]
```diff
@@ -55,3 +55,20 @@ export type LanguageDescriptor<TTextPreprocessorDescriptor extends TextPreproces
...
+export type RemoveDiacriticsPreprocessors = {
+    removeDiacritics: TextPreprocessor<boolean>;
+};
+
+export type CapitalizationPreprocessors = {
+    capitalizeFirstLetter: TextPreprocessor<boolean>;
+    decapitalize: TextPreprocessor<boolean>;
+};
+
+export type LanguageWithCapitalization = LanguageDescriptor<CapitalizationPreprocessors>;
```
Comment (StefanVukovic99) [inline types/ext/language.d.ts:72]: This is just to avoid type files for every language / long type annotations.

Comment (Scrub1492) [reply types/ext/language.d.ts:72]: Might be nice to have some type JSDoc.

File: .eslintrc.json (hunk 1 above)
Comment (StefanVukovic99) [reply .eslintrc.json:505]: Yeah, that fixes one of the tests, but `eslint-config.test.js` in `test-code` I can't figure out

File: types/ext/language.d.ts (hunk 2 above)
Comment (StefanVukovic99) [reply types/ext/language.d.ts:72]: You mean like comments in the .d.ts files explaining the types or smth else? the types should already be used in JSDoc annotations 🤔

Comment (Scrub1492) [reply types/ext/language.d.ts:72]: Yeah, I meant comments in the .d.ts files.

File: ext/js/language/zh/language-chinese.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+
+/** @type {import('language').LanguageWithoutPreprocessors} */
+export const descriptor = {
```
[the remark this answers is not in the document]
Comment (StefanVukovic99) [reply ext/js/language/zh/language-chinese.js:20]: This adds no features for chinese yet, just an option and the example Text. I guess scanning pinyin might work if a dict has pinyin as readings

File: types/ext/language.d.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+export type LanguageSummary = {
+    name: string;
+    iso: string;
+    exampleText: string;
+};
+
+export type LanguageDescriptor<TTextPreprocessorDescriptor extends TextPreprocessorDescriptor> = {
+    name: string;
+    iso: string;
+    exampleText: string;
```
[the remark this answers is not in the document]
Comment (StefanVukovic99) [reply types/ext/language.d.ts:51]: Yeah, like 読め for japanese now. Should be used in my next PR.

File: ext/js/language/languages.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+import {descriptor as descriptorRomanian} from './ro/language-romanian.js';
+import {descriptor as descriptorRussian} from './ru/language-russian.js';
+import {descriptor as descriptorSerboCroatian} from './sh/language-serbo-croatian.js';
+import {descriptor as descriptorAlbanian} from './sq/language-albanian.js';
+import {descriptor as descriptorSwedish} from './sv/language-swedish.js';
+import {descriptor as descriptorThai} from './th/language-thai.js';
+import {descriptor as descriptorVietnamese} from './vi/language-vietnamese.js';
+import {descriptor as descriptorChinese} from './zh/language-chinese.js';
+
+const languageDescriptors = [
```
[the remark this answers is not in the document]
Comment (StefanVukovic99) [reply ext/js/language/languages.js:43]: see https://github.com/themoeway/yomitan/pull/584#issue-2104241060
i'd prefer dynamic, but for now i think there is not too much load on performance/memory doing it this way

**Comment by djahandarie** [comment]
Don't know if we need to do it as part of this PR, but we will probably eventually want to add to CODEOWNERS separate lines for each per-language folder, as individual language concerns will likely need to be reviewed by someone who knows something about that language.

File: ext/js/language/ar/language-arabic.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+import {basicTextPreprocessorOptions} from '../text-preprocessors.js';
+
+/** @type {import('language/language-arabic').ArabicLanguageDescriptor} */
+export const descriptor = {
+    name: 'Arabic',
+    iso: 'ar',
+    exampleText: 'قَرَأَ',
+    textPreprocessors: {
+        removeDiacritics: {
```
Comment (djahandarie) [inline ext/js/language/ar/language-arabic.js:26]: This can likely also be used for Farsi so maybe it should be moved to ../text-preprocessors.js?

**Comment by toasted-nutbread** [comment]
So I haven't gotten into this a whole lot yet, but one thing stand out to me: the amount of tiny files added for language descriptors. Would it be better if we just put these into a single file? IMO the huge bloat of the copyright in each file, followed by like 4 lines of actual code, seems like a big waste. And reducing ~30 files into 1 may just be better for the way that they are imported as well. It's already impossible to dynamically import them anyways, so I don't think we would be losing anything.

What do you and others think?

**Comment by StefanVukovic99** [comment]
I've had the thought that some files were way too big, never too small lol. Having everything language-specific in its own folder seems tidy. I'm quite incredulous that the copyright is needed in every file, and not sure that dynamic loading is absolutely impossible. 

That said, I don't really mind either way, I guess we can do tiny languages in one file (directly in languages.js?), and separate them out when they get to a reasonable size.

**Comment by toasted-nutbread** [comment]
With regard to the header, I mean it's probably not strictly needed if we decide we want to change it, so long as it exists somewhere. It could probably be turned into a minimal stub if there's desire to keep a header that references the full license externally. During the Yomichan era, this thought never really came to mind because all the files were significantly larger than the header itself. Obviously that has been changing.

With regard to files that are too large, I also agree with this on a lot of things and I am doing some work to address this where possible. Unfortunately, there are a lot of complex classes that, at least presently, can't really be broken down easily. `Backend` is a huge cause of issues on that front.

With regard to dynamic loading of files, I agree there are probably some clever ways to handle this, but given the small size and repetitive content of all of the existing and new language descriptors, dynamic loading would probably just result in more overhead at the end of the day.

My two cents is that grouping them all into a unified file would actually make it easier to read through all of the languages with no difference in existing functionality. I can try to mock this up more formally to see how I/we feel about it rather than speculating.

**Comment by toasted-nutbread** [comment]
Language-specific things can still be located in the folders, I just think the final composition of all the descriptors is simplified if it's all declared in a single config file. #704 shows this, while still language-specific things like `japanese-text-preprocessors.js` in the specific language folder.

The reason why I think the single location hosting all the config is useful is because it gives you an overview of all of the languages at a glance without having to open 30+ different files if we need to add a new field to a `LanguageDescriptor`. I'm thinking of it similar to how eslintrc and similar config files can also be written directly in JavaScript.

(At the very least, #704 tidies some of the types, even if we decide we want to have language-specific folders for things.)

**Comment by djahandarie** [comment]
@StefanVukovic99 I think this is good to merge whenever you have time to resolve comments and rebase on top of toasted-nutbread's new file layout

File: types/ext/language.d.ts (hunk 2 above)
Comment (StefanVukovic99) [reply types/ext/language.d.ts]: I've added just a comment on TextPreprocessor, to make up for the now deleted text in settings.html. The rest is hopefully understandable enough without comments.

**Comment by StefanVukovic99** [comment]
I think @toasted-nutbread will want at least one more pass at this.

**Review by toasted-nutbread — APPROVED** [review]
Looks good!

**Review by djahandarie — APPROVED** [review]

**Pull request merged**